internal/service/ec2/vpc_ipam_preview_next_cidr.go

```diff
@@ -0,0 +1,129 @@
+		// },
+		Schema: map[string]*schema.Schema{
+			"cidr": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			// // temp comment out till bug is resolved
+			// "disallowed_cidrs": {
+			// 	Type:     schema.TypeSet,
+			// 	Optional: true,
+			// 	ForceNew: true,
+			// 	Elem: &schema.Schema{
+			// 		Type: schema.TypeString,
+			// 		ValidateFunc: validation.Any(
+			// 			verify.ValidIPv4CIDRNetworkAddress,
+			// 			// Follow the numbers used for netmask_length
+			// 			validation.IsCIDRNetwork(0, 32),
+			// 		),
+			// 	},
+			// },
+			"id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"ipam_pool_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"netmask_length": {
+				// Possible netmask lengths for IPv4 addresses are 0 - 32.
+				// AllocateIpamPoolCidr API
+				//   - If there is no DefaultNetmaskLength allocation rule set on the pool,
+				//   you must specify either the NetmaskLength or the CIDR.
+				//   - If the DefaultNetmaskLength allocation rule is set on the pool,
+				//   you can specify either the NetmaskLength or the CIDR and the
+				//   DefaultNetmaskLength allocation rule will be ignored.
+				// since there is no attribute to check if the rule is set, this attribute is required
+				Type:         schema.TypeInt,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IntBetween(0, 32),
+			},
+		},
+	}
+}
+
+func ResourceVPCIpamPreviewNextCidrCreate(d *schema.ResourceData, meta interface{}) error {
+	poolId := d.Get("ipam_pool_id").(string)
+	uniqueValue := resource.UniqueId()
+
+	// preview will not produce an IpamPoolAllocationId. Hence use the uniqueValue instead
+	d.SetId(encodeVPCIpamPreviewNextCidrID(uniqueValue, poolId))
+
+	return ResourceVPCIpamPreviewNextCidrRead(d, meta)
+}
+
+func ResourceVPCIpamPreviewNextCidrRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).EC2Conn
+
+	uniqueValue, poolId, err := decodeVPCIpamPreviewNextCidrID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	input := &ec2.AllocateIpamPoolCidrInput{
+		ClientToken:     aws.String(uniqueValue),
+		IpamPoolId:      aws.String(poolId),
+		NetmaskLength:   aws.Int64(int64(d.Get("netmask_length").(int))),
+		PreviewNextCidr: aws.Bool(true),
+	}
+
+	// if v, ok := d.GetOk("disallowed_cidrs"); ok && v.(*schema.Set).Len() > 0 {
+	// 	input.DisallowedCidrs = flex.ExpandStringSet(v.(*schema.Set))
+	// }
+
+	output, err := conn.AllocateIpamPoolCidr(input)
+
+	if err != nil {
+		return fmt.Errorf("Error allocating cidr from IPAM pool (%s): %w", d.Get("ipam_pool_id").(string), err)
+	}
+
+	if output == nil || output.IpamPoolAllocation == nil {
+		return fmt.Errorf("error allocating from ipam pool (%s): empty response", poolId)
+	}
+
+	d.Set("cidr", output.IpamPoolAllocation.Cidr)
+
+	return nil
+}
+
+func encodeVPCIpamPreviewNextCidrID(uniqueValue, poolId string) string {
```

Comment:
Could we just export the allocation ID functions and use those instead?

https://github.com/hashicorp/terraform-provider-aws/blob/f445f5e1980a8b4a6581b2964b8ba70b80330128/internal/service/ec2/vpc_ipam_pool_cidr_allocation.go#L194-L204

Reply:
I wanted to initially but the name of the first parameter is `allocation_id ` which deviates from how I used `resource.UniqueId()`. I was toying with the idea of generalising the function in the file you mentioned above. Should I rename the parameter in that function and reuse it in this case?

Reply:
I wonder if there is opportunity here to move these functions to an even more general location. im sure lots of resources construct a custom ID with 2 strings.

Reply:
Resolved in 4d7cce44fba5581e493b4ad64e4f64f382ba6725